### ISSUE TITLE
Script generator: text in fields disappering as it is typed

### ISSUE THREAD

**Issue by Tom-Willemsen** [ISISComputingGroup/IBEX#5708]
As a script generator user I would like to be able to type things into the input fields without them being deleted.

A video of this bug is available at `isis\shares\isis_experiment_controls\data for tickets\Ticket5708_script_gen_deleting_input`

**Comment by John-Holt-Tessella** [comment]
I think I can reproduce this on my machine by creating a long script 20+ lines long (depending on the speed of your machine)
Quickly add a dozen new actions, then type in one.
It looks like each validate reselects the contents of the box you are in, and thus the next keepstroke deletes the contents of the box.
(Although it looks like on the video that the lines are also failing to validate)

**Comment by Tom-Willemsen** [comment]
The basic problem is that `setRows` is called multiple times, with the number of times it is called being proportional to the number of rows in the table.

Each time `setRows` is called it implicitly puts the table into non-editing mode briefly, and then back into editing mode once the call completes.

Each time the table goes back into "editing mode", it highlights the current content of the cell, which means that the next keystroke will clear the content of the cell.

I tried selectively setting the cell into editing mode only if (not editing) and (row/column isn't already selected), but this then has the effect that the rows don't end up in edit mode at all, meaning that you have to laboriously click in each cell that you want to type into.

Need to test if I can get into the `EditingSupport` classes in some way to find whether an edit is in progress when calling `setRows`. Or perhaps cache editing state from before calling `setRows` and then use that cached state in the decision to reactivate editor.

**Comment by John-Holt-Tessella** [comment]
Estimated at 3 days but could take longer, feel free to reestimate.

**Comment by John-Holt-Tessella** [comment]
Thought: Is the update/validation happening on edit instead of on move cell?

**Comment by Tom-Willemsen** [comment]
I think the PR above covers one of the cases of this error. I can still reproduce it (briefly) by adding lots of rows at a time but I think the PR above makes it better.

**Comment by DominicOram** [comment]
I've done a hotfix that reapplies the selection when you move focus at https://github.com/ISISComputingGroup/ibex_gui/pull/1193.

Tom's description of the underlying cause was close but not quite correct. What actually happens is this:
1) The user edits a cell and tabs away
2) The backend sees the cell has been edited and recalculates the validation and time estimate
3) The backend calls `setRows` 3 times, for the new value, the new validation and the new time estimate
4) `setRows` calls `setInput`, this causes the focus of the cell to be lost
5) `setRows` then sets the focus back on the cell, selecting the whole cell
6) When the user types now it replaces anything in the cell

In the case where there are fewer rows this happens quickly enough so that the user doesn't notice it and actually selecting the whole cell when it is being first tabbed into is correct. However, when there are more rows step 2 takes longer and so the user has started entering a new value by the time step 6 occurs. This is then exacerbated by the fact that step 4 is interpreted as the user entering a new value and so the process goes back to step 1 and `setRows` is called another 3 times. If the user waits sufficiently long between tabbing away and entering a new value this is fine.

The hotfix follows the same mechanism but after step 5 it will work out where the user had previously selected and reselect there rather than selecting the whole cell. This is a temporary fix for the symtpom and won't fix the underlying issue (as Tom's reproduction proves). 

I believe the correct fix is to not call `setInput` each time that the data in the table updates. `setInput` should provide the table with an observable list, which the table can monitor for any changes. I think that when the observable list is updated it will not cause the loss of focus.

**Comment by kjwoodsISIS** [comment]
If the hotfix is now implemented, what remains to be done on this ticket?
Presumably a long-term solution will be the subject of another ticket.

**Comment by DominicOram** [comment]
I've pointed this to provide the long term solution and it was the next thing I was going to work on.

**Comment by DominicOram** [comment]
Upon reflection I agree with @kjwoodsISIS that a proper fix should be done in a separate ticket. To review this:
* Confirm original issue is harder to reproduce (I think @Tom-Willemsen already did this)
* Confirm there is detail for how to move forward
